README.md
# SubgRfspy

Elixir SubgRfspy is a package to communicate with TI cc111x chips running the wireless packet
firmware [subg_rfspy](https://github.com/ps2/subg_rfspy). UART and SPI are the two serial protocols supported.

## Installation

Add `subg_rfspy` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:subg_rfspy, "~> 1.0.0"}
  ]
end
```


CHANGELOG.md
## v2.0.0

This release introduces several features and breaks compatibility with 1.x. There is now a struct for the UART and SPI chips that holds the relevant information for each. start_link should now be called with the struct. start_link will alias the process with the atom given in the name key of the struct, so it's possible to start more than one chip process at one time now.

When calling any of the public functions on the SubgRfspy module, the first parameter is the UART or SPI struct that was used to call start_link. This allows the SubgRfspy module to use the name key of the struct to communicate with the process, and also allows upstream libraries to more easily define a protocol for the chip functions using the structs.

Additionally, a new chip_present? function has been added to UART and SPI that will attempt to detect whether the chip is responding using the given struct's pin and device configuration.

In some ways the library has gotten more complex because of these changes, but it solves several longstanding problems with supporting different kinds of serial devices and chips, and this should pave the way to making some simplifications in the design for developers using this package as the simplifications become clearer.


## v1.0.0

This release marks the feature complete, minimum viable state of the package. In addition to UART, SPI is now supported.

## v0.9.0

This is an initial release of the package, containing the following:

* Support for communication with a subg_rfspy chip via a UART and the Nerves.UART
* Support for recording UART communication to a csv file for analysis and playback in tests


config/test.exs
use Mix.Config

config :logger, level: :info
config :subg_rfspy, :serial_driver, SubgRfspy.UARTProxy


test/subg_rfspy_test.exs
defmodule SubgRfspyTest do
  use ExUnit.Case, async: false

  alias SubgRfspy.UARTProxy

  doctest SubgRfspy

  defp assert_down(pid) do
    ref = Process.monitor(pid)
    assert_receive {:DOWN, ^ref, _, _, _}
  end

  setup(context) do
    {:ok, pid} = UARTProxy.start_link(context[:test])
    UARTProxy.record
    on_exit(fn() -> assert_down(pid) end)
    :ok
  end

  @write_register "06"
  test "set base frequency writes correct register data" do
    SubgRfspy.set_base_frequency(916.500)

    # freq0 (0x09) should have the upper frequency byte of 0x26
    assert Enum.member?(UARTProxy.interactions, ["write", @write_register <> "0926", "ok"])
    # freq1 (0x0A) should have the middle frequency byte of 0x30
    assert Enum.member?(UARTProxy.interactions, ["write", @write_register <> "0A30", "ok"])
    # freq2 (0x0B) should have the lower frequency byte of 0x00
    assert Enum.member?(UARTProxy.interactions, ["write", @write_register <> "0B00", "ok"])
  end

  test "sync interacts with the chip firmware correctly" do
    SubgRfspy.sync()

    assert Enum.member?(UARTProxy.interactions, ["write", "01", "ok"])
    assert Enum.member?(UARTProxy.interactions, ["write", "02", "ok"])
  end
end


mix.exs
defmodule SubgRfspy.Mixfile do
  use Mix.Project

  def project do
    [
      app: :subg_rfspy,
      version: "2.0.0",
      elixir: "~> 1.5",
      start_permanent: Mix.env == :prod,
      deps: deps(),
      name: "SubgRfspy",
      source_url: "https://github.com/infinity-aps/elixir_subg_rfspy",
      description: description(),
      package: package()
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [{:nerves_uart, "~> 0.1.1"},
     {:elixir_ale, "~> 1.0.1"},
     {:csv, "~> 2.0.0"},
     {:credo, "~> 0.8", only: [:dev, :test], runtime: false},
     {:ex_doc, ">= 0.0.0", only: :dev}]
  end

  defp description do
    """
    SubgRfspy is a library to handle sub-GHz wireless packet communication via a TI cc1110 chip running subg_rfspy.
    """
  end

  defp package do
    [
      maintainers: ["Timothy Mecklem"],
      licenses: ["MIT License"],
      links: %{"Github" => "https://github.com/infinity-aps/elixir_subg_rfspy"}
    ]
  end
end


lib/subg_rfspy/spi.ex
defmodule SubgRfspy.SPI do
  @moduledoc """
  This GenServer provides the most basic read and write operations to communicate with a wireless chip running
  subg_rfspy firmware (https://github.com/ps2/subg_rfspy) over SPI.
  """

  defstruct name: nil, device: nil, reset_pin: nil

  require Logger
  use GenServer
  alias ElixirALE.{GPIO, SPI}

  @initial_byte 0x99

  def start_link(%SubgRfspy.SPI{name: name, device: device, reset_pin: reset_pin}) do
    GenServer.start_link(__MODULE__, [device, reset_pin], name: name)
  end

  @status_tx1 <<@initial_byte::8, 0x010100::24>>
  @status_tx2 <<@initial_byte::8, 0x00000000::32>>
  def init([device, reset_pin]) do
    with {:ok, serial_pid} <- SPI.start_link(device, [speed_hz: 62500]),
         {:ok, reset_pid} <- GPIO.start_link(reset_pin, :output),
         :ok <- _reset(reset_pid),
         transfer(serial_pid, @status_tx1),
         <<_::16>> <> "OK" <> <<_::8>> <- transfer(serial_pid, @status_tx2) do
      {:ok, %{serial_pid: serial_pid, reset_pid: reset_pid, read_queue: []}}
    else
      error ->
        Logger.error fn -> "The SPI failed to start: #{inspect(error)}" end
        {:error, "The SPI failed to start"}
    end
  end

  def chip_present?(%SubgRfspy.SPI{device: device, reset_pin: reset_pin}) do
    with {:ok, serial_pid} <- SPI.start_link(device, [speed_hz: 62500]),
         {:ok, reset_pid} <- GPIO.start_link(reset_pin, :output),
         :ok <- _reset(reset_pid),
         transfer(serial_pid, @status_tx1),
         <<_::16>> <> "OK" <> <<_::8>> <- transfer(serial_pid, @status_tx2) do

      SPI.release(serial_pid)
      GPIO.release(reset_pid)
      true
    else
      _error -> false
    end
  end

  def handle_call({:clear_buffers}, _from, state) do
    {:reply, :ok, state}
  end

  def handle_call({:read, _timeout_ms}, _from, state = %{read_queue: [head | rest]}) do
    {:reply, {:ok, head}, Map.put(state, :read_queue, rest)}
  end

  def handle_call({:read, timeout_ms}, _from, state = %{serial_pid: serial_pid}) do
    timeout_time = System.monotonic_time(:millisecond) + timeout_ms + 5_000
    {:reply, _read(serial_pid, timeout_time), state}
  end

  def handle_call({:write, data, _timeout_ms}, _from, state = %{serial_pid: serial_pid}) do
    length_of_tx = byte_size(data)
    <<_::8, length_of_rx::8>> = transfer(serial_pid, <<@initial_byte::8, length_of_tx::8>>)
    extra_bits = (max(length_of_tx, length_of_rx) - length_of_tx) * 8
    tx_bytes = case extra_bits do
                 0 -> data
                 x -> data <> <<0x00::size(x)>>
               end
    rx = transfer(serial_pid, tx_bytes)
    state = case length_of_rx do
      0 -> state
      n ->
        bytes_to_read = n - 1
        <<real_rx::binary-size(bytes_to_read), _::binary>> = rx
        Map.put(state, :read_queue, state.read_queue ++ String.split(real_rx, <<0x00>>))
    end
    {:reply, :ok, state}
  end

  def handle_call({:reset}, _from, state = %{reset_pid: reset_pid}) do
    {:reply, _reset(reset_pid), state}
  end

  defp _reset(reset_pid) do
    GPIO.write(reset_pid, 0)
    :timer.sleep(10)
    GPIO.write(reset_pid, 1)
    :timer.sleep(2001)
    :ok
  end

  def terminate(reason, %{serial_pid: serial_pid}) do
    Logger.warn fn -> "Exiting, reason: #{inspect reason}" end
    SPI.release(serial_pid)
  end

  defp _read(pid, timeout_time) do
    case System.monotonic_time(:millisecond) do
      x when x > timeout_time -> {:error, :timeout}
      _ ->
        <<_::8, length_of_rx::8>> = transfer(pid, <<@initial_byte::8, 0x00::8>>)
        case length_of_rx do
          0 ->
            :timer.sleep(100)
            _read(pid, timeout_time)
          _ ->
            extra_bits = length_of_rx * 8
            tx_bytes = <<0x00::size(extra_bits)>>
            rx_byte_count = length_of_rx - 1
            <<rx::binary-size(rx_byte_count), _::8>> = transfer(pid, tx_bytes)
            {:ok, rx}
        end
    end
  end

  defp transfer(pid, data) do
    tx = data |> reverse_bits()
    rx = pid |> SPI.transfer(tx) |> reverse_bits()
    Logger.debug fn() -> "Sent: #{Base.encode16(data)} Received: #{Base.encode16(rx)}" end
    rx
  end

  defp reverse_bits(data), do: _reverse_bits(data, <<>>)
  defp _reverse_bits(<<>>, reversed), do: reversed
  defp _reverse_bits(<<first::binary-size(1), rest::binary>>, reversed) do
    <<a::1, b::1, c::1, d::1, e::1, f::1, g::1, h::1>> = first
    _reverse_bits(rest, <<reversed::binary, h::1, g::1, f::1, e::1, d::1, c::1, b::1, a::1>>)
  end
end


lib/subg_rfspy/uart.ex
defmodule SubgRfspy.UART do
  @moduledoc """
  This GenServer provides the most basic read and write operations to communicate with a wireless chip running
  subg_rfspy firmware (https://github.com/ps2/subg_rfspy) over a UART.
  """

  defstruct name: nil, device: nil

  require Logger
  use GenServer
  alias SubgRfspy.UARTFraming
  alias Nerves.UART

  def start_link(%SubgRfspy.UART{name: name, device: device}) do
    GenServer.start_link(__MODULE__, device, name: name)
  end

  @reset 0x07
  def init(device) do
    with {:ok, serial_pid} <- UART.start_link,
         :ok <- UART.open(serial_pid, device, speed: 19_200, active: false),
         :ok <- UART.configure(serial_pid, framing: {UARTFraming, separator: <<0x00>>}),
         :ok <- UART.flush(serial_pid) do

      {:ok, serial_pid}
    else
      error ->
        Logger.error fn -> "The UART failed to start: #{inspect(error)}" end
        {:error, "The UART failed to start"}
    end
  end

  def chip_present?(%SubgRfspy.UART{device: device}) do
    with {:ok, serial_pid} <- UART.start_link,
         :ok <- UART.open(serial_pid, device, speed: 19_200, active: false),
         :ok <- UART.configure(serial_pid, framing: {UARTFraming, separator: <<0x00>>}),
         :ok <- UART.flush(serial_pid),
         :ok <- write_fully(<<@reset>>, 100, serial_pid),
         :timer.sleep(2000),
         :ok <- write_fully(<<0x01>>, 100, serial_pid),
         {:ok, "OK"} <- UART.read(serial_pid, 2_000) do

      UART.stop(serial_pid)
      true
    else
      _error -> false
    end
  end

  def terminate(reason, serial_pid) do
    Logger.warn fn -> "Exiting, reason: #{inspect reason}" end
    UART.close(serial_pid)
  end

  def handle_call({:write, data, timeout_ms}, _from, serial_pid) do
    {:reply, write_fully(data, timeout_ms, serial_pid), serial_pid}
  end

  def handle_call({:read, timeout_ms}, _from, serial_pid) do
    # is_uart_running = "ps" |> System.cmd([]) |> elem(0) |> String.contains?("uart")
    # if !is_uart_running do
    #   Logger.debug fn -> "UART port is not running!" end
    # end
    {:reply, UART.read(serial_pid, timeout_ms + 1_000), serial_pid}
  end

  def handle_call({:clear_buffers}, _from, serial_pid) do
    {:reply, UART.flush(serial_pid), serial_pid}
  end

  defp write_fully(data, timeout_ms, serial_pid) do
    case UART.write(serial_pid, data, timeout_ms) do
      :ok -> UART.drain(serial_pid)
      err -> err
    end
  end
end


lib/subg_rfspy/uart_proxy.ex
defmodule SubgRfspy.UARTProxy do
  @moduledoc """
  UARTProxy is a module that can sit between the UART module and a caller and capture the traffic between the two for
  later playback in tests. It is intended to be used as a drop in replacement for UART during configuration.
  """

  use GenServer
  alias ExUnit.Assertions
  alias SubgRfspy.UART
  require Logger

  @genserver_timeout 60_000

  def start_link(context_name) do
    File.mkdir_p("test/cassettes")
    record = System.get_env("RECORD_CASSETTE") || "false"
    initial_state = case record do
      "true" ->
        _start_link(context_name, true)
      "new" ->
        cassette_exists = File.exists?(cassette_filename(context_name))
        _start_link(context_name, !cassette_exists)
      "false" ->
        _start_link(context_name, false)
    end

    GenServer.start_link(__MODULE__, initial_state, name: __MODULE__)
  end

  def _start_link(context_name, true) do
    Logger.debug fn ->
      "Starting #{context_name} in record mode"
    end

    device = System.get_env("SUBG_RFSPY_DEVICE")
    case device do
      nil ->
        message = "SUBG_RFSPY_DEVICE must be set for record mode"
        Logger.error fn -> message end
        {:error, message}
      _ ->
        {:ok, _} = UART.start_link(device)
        %{record: true, interactions: [], context_name: context_name, recording: false}
    end
  end

  def _start_link(context_name, false) do
    Logger.debug fn -> "Starting #{context_name} in playback mode" end
    filename = context_name |> cassette_filename()
    expected_interactions = case File.exists?(filename) do
                              true -> filename |> File.stream!() |> CSV.decode! |> Enum.map(&(&1))
                              false -> []
                            end
    %{record: false, interactions: [], remaining_interactions: expected_interactions, recording: false}
  end

  def write(data, timeout_ms) do
    GenServer.call(__MODULE__, {:write, data, timeout_ms}, @genserver_timeout)
  end

  def read(timeout_ms) do
    GenServer.call(__MODULE__, {:read, timeout_ms}, @genserver_timeout)
  end

  def clear_buffers do
    GenServer.call(__MODULE__, {:clear_buffers}, @genserver_timeout)
  end

  def record do
    GenServer.call(__MODULE__, {:record}, @genserver_timeout)
  end

  def pause do
    GenServer.call(__MODULE__, {:pause}, @genserver_timeout)
  end

  def interactions do
    GenServer.call(__MODULE__, :interactions, @genserver_timeout)
  end

  # GENSERVER LIFECYCLE

  def init(initial_state) do
    Process.flag :trap_exit, true
    {:ok, initial_state}
  end

  def terminate(_, %{record: true, interactions: []}), do: nil
  def terminate(_, %{record: true, interactions: interactions, context_name: context_name}) do
    file = File.open!(cassette_filename(context_name), [:write, :utf8])
    interactions |> Enum.reverse |> CSV.encode |> Enum.each(&IO.write(file, &1))
  end

  def terminate(_, %{interactions: expected_interactions}) do
    Assertions.assert expected_interactions, []
  end

  # PLAYBACK MODE

  def handle_call({:write, data, _timeout_ms}, _from, state = %{record: false}) do
    %{remaining_interactions: [["write", expected_data, response] | rest], interactions: interactions} = state
    new_interactions = [["write", expected_data, response] | interactions]
    {:ok, expected_data} = Base.decode16(expected_data)
    Assertions.assert expected_data, data
    new_state = %{state | interactions: new_interactions, remaining_interactions: rest}
    {:reply, String.to_atom(response), new_state}
  end

  def handle_call({:read, _timeout_ms}, _from, state = %{record: false}) do
    %{remaining_interactions: [["read", expected_data, response] | rest], interactions: interactions} = state
    new_interactions = [["read", expected_data, response] | interactions]
    {:ok, expected_data} = Base.decode16(expected_data)
    new_state = %{state | interactions: new_interactions, remaining_interactions: rest}
    {:reply, {String.to_atom(response), expected_data}, new_state}
  end

  def handle_call({:clear_buffers}, _from, state = %{record: false}) do
    {:reply, :ok, state}
  end

  # RECORD MODE

  def handle_call({:write, data, timeout_ms}, _from, state = %{record: true, interactions: interactions}) do
    response = UART.write(data, timeout_ms)
    state = case state.recording do
              true ->
                interaction = ["write", Base.encode16(data), Atom.to_string(response)]
                %{state | interactions: [interaction | interactions]}
              false -> state
            end
    {:reply, response, state}
  end

  def handle_call({:read, timeout_ms}, _from, state = %{record: true, interactions: interactions}) do
    {response, data} = UART.read(timeout_ms)
    Logger.debug fn -> "Received response from Real UART: {#{response}, #{data}}" end
    state = case state.recording do
              true ->
                interaction = ["read", Base.encode16(data), Atom.to_string(response)]
                %{state | interactions: [interaction | interactions]}
              false -> state
            end
    {:reply, {response, data}, state}
  end

  def handle_call({:clear_buffers}, _from, state = %{record: true}) do
    {:reply, UART.clear_buffers(), state}
  end

  def handle_call({:record}, _from, state) do
    {:reply, :ok, %{state | recording: true}}
  end

  def handle_call({:pause}, _from, state) do
    {:reply, :ok, %{state | recording: false}}
  end

  def handle_call(:interactions, _from, state = %{interactions: interactions}) do
    {:reply, interactions |> Enum.reverse, state}
  end

  defp cassette_filename(context_name) do
    "test/cassettes/#{context_name |> Atom.to_string() |> String.replace(" ", "_")}.csv"
  end
end


lib/subg_rfspy/uart_framing.ex
defmodule SubgRfspy.UARTFraming do
  @behaviour Nerves.UART.Framing

  defmodule State do
    @moduledoc false
    defstruct [
      max_length: nil,
      separator: nil,
      processed: <<>>,
      in_process: <<>>
    ]
  end

  def init(args) do
    max_length = Keyword.get(args, :max_length, 4096)
    separator = Keyword.get(args, :separator, "\n")

    state = %State{max_length: max_length, separator: separator}
    {:ok, state}
  end

  def add_framing(data, state) do
    {:ok, data, state}
  end

  def remove_framing(data, state) do
    {new_processed, new_in_process, lines} =
      process_data(state.separator,
                  byte_size(state.separator),
                  state.max_length,
                  state.processed,
                  state.in_process <> data, [])
    new_state = %{state | processed: new_processed, in_process: new_in_process}
    rc = if buffer_empty?(new_state), do: :ok, else: :in_frame
    {rc, lines, new_state}
  end

  def frame_timeout(state) do
    partial_line = {:partial, state.processed <> state.in_process}
    new_state = %{state | processed: <<>>, in_process: <<>>}
    {:ok, [partial_line], new_state}
  end

  def flush(direction, state) when direction == :receive or direction == :both do
    %{state | processed: <<>>, in_process: <<>>}
  end
  def flush(_direction, state) do
    state
  end

  def buffer_empty?(state) do
    state.processed == <<>> and state.in_process == <<>>
  end

  # Handle not enough data case
  defp process_data(_separator, sep_length, _max_length, processed, to_process, lines)
                    when byte_size(to_process) < sep_length do
    {processed, to_process, lines}
  end

  # Process data until separator or next char
  defp process_data(separator, sep_length, max_length, processed, to_process, lines) do
    case to_process do
      # Handle separator
      <<^separator::binary-size(sep_length), rest::binary>> ->
        new_lines = lines ++ [processed]
        process_data(separator, sep_length, max_length, <<>>, rest, new_lines)
      # Handle line too long case
      to_process when byte_size(processed) == max_length and to_process != <<>> ->
        new_lines = lines ++ [{:partial, processed}]
        process_data(separator, sep_length, max_length, <<>>, to_process, new_lines)
      # Handle next char
      <<next_char::binary-size(1), rest::binary>> ->
        process_data(separator, sep_length, max_length, processed <> next_char, rest, lines)
    end
  end

end


lib/subg_rfspy.ex
defmodule SubgRfspy do
  @moduledoc """
  This module models the serial communications with a chip running the subg_rfspy firmware
  (https://github.com/ps2/subg_rfspy). The lower level serial communication is handled by a driver such as
  SubgRfspy.UART.
  """

  require Logger
  use Bitwise

  @serial_timeout_ms_padding 1000

  @channel 0
  @retry_count 0
  @repetitions 1
  @repetition_delay 0

  @commands %{
    get_state:       0x01,
    get_version:     0x02,
    get_packet:      0x03,
    send_packet:     0x04,
    send_and_listen: 0x05,
    update_register: 0x06,
    reset:           0x07
  }

  @registers %{
    freq2:    0x09,
    freq1:    0x0A,
    freq0:    0x0B,
    mdmcfg4:  0x0C,
    mdmcfg3:  0x0D,
    mdmcfg2:  0x0E,
    mdmcfg1:  0x0F,
    mdmcfg0:  0x10,
    agcctrl2: 0x17,
    agcctrl1: 0x18,
    agcctrl0: 0x19,
    frend1:   0x1A,
    frend0:   0x1B
  }

  def update_register(%{name: name}, register, value) do
    GenServer.call(name, {:clear_buffers})
    write_command(name, <<register::8, value::8>>, :update_register, 100)

    true = read_until(name, <<1>>, 5)
  end

  def set_base_frequency(chip, mhz) do
    freq_xtal = 24_000_000
    val = round((mhz * 1_000_000) / (freq_xtal / :math.pow(2, 16)))
    update_register(chip, @registers[:freq0], val &&& 0xff)
    update_register(chip, @registers[:freq1], (val >>> 8) &&& 0xff)
    update_register(chip, @registers[:freq2], (val >>> 16) &&& 0xff)
    {:ok}
  end

  def read(%{name: name}, timeout_ms \\ 1000) do
    write_command(name, <<@channel::8, timeout_ms::32>>, :get_packet, timeout_ms + 1000)
    name |> read_response(timeout_ms) |> process_response()
  end

  def write(%{name: name}, packet, repetitions, repetition_delay, timeout_ms) do
    write_batches(name, packet, repetitions, repetition_delay, timeout_ms)
  end

  def write_and_read(%{name: name}, packet, timeout_ms \\ 500) do
    command = <<@channel::8, @repetitions::8, @repetition_delay::8,
      @channel::8, timeout_ms::size(32), @retry_count::8,
      packet::binary>>
    write_command(name, command, :send_and_listen, timeout_ms + @serial_timeout_ms_padding)
    padded_timeout = timeout_ms + @serial_timeout_ms_padding
    name |> read_response(padded_timeout) |> process_response()
  end

  def reset(%{name: name}) do
    :ok = write_command(name, <<>>, :reset, 100)
  end

  def sync(name) do
    GenServer.call(name, {:clear_buffers})
    {:ok, status} = get_state(name)
    {:ok, version} = get_version(name)
    %{status: status, version: version}
  end

  def get_version(%{name: name}) do
    :ok = write_command(name, <<>>, :get_version, 100)
    read_response(name, 5000)
  end

  def get_state(%{name: name}) do
    :ok = write_command(name, <<>>, :get_state, 100)
    read_response(name, 5000)
  end

  @max_repetition_batch_size 250
  defp write_batches(name, packet, repetitions, repetition_delay, timeout_ms) do
    case repetitions do
      x when x > @max_repetition_batch_size ->
        write_batch(name, packet, repetitions - @max_repetition_batch_size, repetition_delay, timeout_ms)
        read_response(name, timeout_ms)

        write_batches(name, packet, repetitions - @max_repetition_batch_size, repetition_delay, timeout_ms)
      _ ->
        write_batch(name, packet, repetitions, repetition_delay, timeout_ms)
    end
  end

  defp write_batch(name, packet, repetitions, repetition_delay, timeout_ms) do
    command = <<@channel::8, repetitions::8, repetition_delay::8, packet::binary>>
    write_command(name, command, :send_packet, timeout_ms)

    read_response(name, timeout_ms)
  end

  defp write_command(name, param, command_type, timeout_ms) do
    command = @commands[command_type]
    real_timeout = timeout_ms + 10_000
    response = GenServer.call(name, {:write, <<command::8>> <> param, real_timeout}, genserver_timeout(real_timeout))
    if command_type == :reset do
      :timer.sleep(5000)
    end
    response
  end

  @timeout             0xAA
  @command_interrupted 0xBB
  @zero_data           0xCC
  defp read_response(name, timeout_ms) do
    response = GenServer.call(name, {:read, timeout_ms}, genserver_timeout(timeout_ms))
    case response do
      {:ok, <<@command_interrupted>>} ->
        Logger.debug fn -> "Command Interrupted, continuing to read" end
        read_response(name, timeout_ms)
      _ ->
        response
    end
  end

  defp process_response({:error, :timeout}),              do: {:error, :timeout}
  defp process_response({:ok, <<@timeout>>}),             do: {:error, :timeout}
  defp process_response({:ok, <<>>}),                     do: {:error, :empty}
  defp process_response({:ok, <<@command_interrupted>>}), do: {:error, :command_interrupted}
  defp process_response({:ok, <<@zero_data>>}),           do: {:error, :zero_data}
  defp process_response({:ok, <<raw_rssi::8, sequence::8, data::binary>>}) do
    {:ok, %{rssi: rssi(raw_rssi), sequence: sequence, data: data}}
  end

  @rssi_offset 73
  defp rssi(raw_rssi) when raw_rssi >= 128, do: rssi(raw_rssi - 256)
  defp rssi(raw_rssi), do: (raw_rssi / 2) - @rssi_offset

  defp read_until(_, _, 0), do: false
  defp read_until(name, expected, retries) do
    case read_response(name, 100) do
      {:ok, ^expected} -> true
      _                -> read_until(name, expected, retries - 1)
    end
  end

  defp genserver_timeout(timeout_ms), do: timeout_ms + 10_000
end
